subprojects/core/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultProjectDependency.java

```diff
@@ -80,7 +81,7 @@ public Configuration findProjectConfiguration() {
         String declaredConfiguration = getTargetConfiguration();
         Configuration selectedConfiguration = dependencyConfigurations.getByName(GUtil.elvis(declaredConfiguration, Dependency.DEFAULT_CONFIGURATION));
         if (declaredConfiguration!=null && !selectedConfiguration.isCanBeConsumed()) {
```

Comment:
I think this should fail unconditionally now. There might be a missing test case.